packages/app-builder/src/services/data/pivot.tsx

```diff
@@ -0,0 +1,91 @@
+import {
+  type DataModel,
+  type DataModelField,
+  type LinkToSingle,
+  type TableModel,
+} from '@app-builder/models';
+
+interface FieldPivot {
+  type: 'field';
+  baseTableId: string;
+  fieldId: string;
+  id: string;
+  displayValue: string;
+}
+
+interface LinkPivot {
+  type: 'link';
+  baseTableId: string;
+  pathLinkIds: string[];
+  id: string;
+  displayValue: string;
+}
+
+export type Pivot = FieldPivot | LinkPivot;
+
+export function getPivotOptions(
+  tableModel: TableModel,
+  dataModel: DataModel,
+): Pivot[] {
+  const pivots: Pivot[] = [];
+
+  tableModel.fields
+    // Only allow pivots on string fields
+    .filter((field) => field.dataType === 'String')
+    .map((field: DataModelField) => {
+      pivots.push({
+        type: 'field',
+        baseTableId: tableModel.id,
+        fieldId: field.id,
+        id: field.id,
+        displayValue: field.name,
+      });
+    });
+
+  const tablesMap = new Map(dataModel.map((table) => [table.id, table]));
+
+  function recursiveLinkedPivot(
```

Comment:
just a syntax preference here, but I'd suggest making this separately defined recursive function that returns a set/array of possible pivots, rather than a closure defined inside getPivotOptions and mutating the variable `pivots` (I think that makes it easier to read, and we're not really in a case where there is a strong argument for a closure)

Reply:
Done.
I also added some basic tests (but it's not easy to read due to the datamodel + links creation)

Reply:
Concerning "2nd level options", it worked on my side (before and after the refactoring).

Can you check your datamodel "2nd level option" is a String field ?
NB: the pivot naming should be a concatenation of links names

Reply:
seems good now anyway 👍